src/ApacheOrcDotNet.OptimizedReader/BufferReader.cs

```diff
@@ -0,0 +1,52 @@
+﻿using System;
+
+namespace ApacheOrcDotNet.OptimizedReader
+{
+    public readonly ref struct BufferReader
+    {
+        private readonly ReadOnlySpan<byte> _buffer;
+        private readonly int _position;
+
+        public BufferReader(ReadOnlySpan<byte> buffer)
+        {
+            _buffer = buffer;
+            _position = 0;
+        }
+
+        public bool Complete => _position >= _buffer.Length;
+
+        public bool TryRead(out byte value)
+        {
+            value = default;
+
+            if (_position >= _buffer.Length)
+                return false;
+
+            value = _buffer[_position];
+
+            Advance(1);
+
+            return true;
+        }
+
+        public bool TryCopyTo(Span<byte> buffer)
+        {
+            if (_position + buffer.Length > _buffer.Length)
+                return false;
+
+            _buffer.Slice(_position, buffer.Length).CopyTo(buffer);
+
+            Advance(buffer.Length);
+
+            return true;
+        }
+
+        private unsafe void Advance(int length)
+        {
+            fixed (int* ptr = &_position)
```

Comment:
I agree, this is not the most clean implementation.

The main reason I've created this struct was to be able to centralize readings, without having to add `return` or `out` values in Buffer/RLE methods to keep track of the current read `index`. Unfortunately, not having the struct marked as `readonly` will prevent using the TryRead/TryCopy methods from within places where the reader was passed by reference.

The issue happens when we try to copy data into a local `Span<byte>` and the compiler cannot guarantee that the `reader` reference will not try to use its data after the current method unrolls.

I end up following this path because, back then, it looked like an acceptable trade-of, since it simplified a lot the reading tree.

If necessary, this entire `BufferReader` struct can be replaced by the buffer `Span<byte>` and a reference to the current read `index`.

Reply:
I like `BufferReader`.  I think it's clear and helpful.  But this unsafe stuff leaves a bit concerned.

Could you plan to walk me through the issue you're describing where the struct isn't readonly on a call tomorrow?